### ISSUE TITLE
Rule S3416: include constructors and reduce FPs

### ISSUE THREAD

**Issue by costin-zaharia-sonarsource** [SonarSource/sonar-dotnet#8934]
While validating S3416, we noticed some FPs that appear when the logger is built by an external call that injects it into the constructor call. 

Some of these FPs were fixed in https://github.com/SonarSource/sonar-dotnet/pull/8932 (but not all of them).  Additionally, as a result of the validation, the rule was removed from SonarWay.

After release, we should:
- check the rule behavior on SonarCloud. If the precision is good promote it to SonarWay.
- reduced the number of FPs
- update the rule to check also constructor declarations:
```csharp
public class BaseType { }

public class EnclosingType : BaseType
{
    public EnclosingType(ILogger<EnclosingType> logger) // Compliant
    {
    }

    public EnclosingType(ILogger<string> logger)                                // Noncompliant {{Update this logger to use its enclosing type.}}
    {
    }

    public EnclosingType(ILogger<BaseType> logger)                              // Noncompliant {{Update this logger to use its enclosing type.}}
    {
    }
}

public class Factory
{
    public EnclosingType CreateType(ILogger<string> logger) => null;            // Compliant
}
```
